src/utility/svd.jl

```diff
@@ -0,0 +1,240 @@
+using TensorKit:
+    SectorDict,
+    _tsvd!,
+    _empty_svdtensors,
+    _compute_svddata!,
+    _create_svdtensors,
+    NoTruncation,
+    TruncationSpace
+CRCExt = Base.get_extension(KrylovKit, :KrylovKitChainRulesCoreExt)
+
+"""
+    PEPSKit.tsvd(t::AbstractTensorMap, alg; trunc=notrunc(), p=2)
+
+Wrapper around `TensorKit.tsvd` which dispatches on the `alg` argument.
+This is needed since a custom adjoint for `PEPSKit.tsvd` may be defined,
+depending on the algorithm. E.g., for `IterSVD` the adjoint for a truncated
+SVD from `KrylovKit.svdsolve` is used.
+"""
+function PEPSKit.tsvd(
+    t::AbstractTensorMap, alg; trunc::TruncationScheme=notrunc(), p::Real=2
+)
+    return TensorKit.tsvd(t; alg, trunc, p)
+end
+
+# Wrapper around Krylov Kit's GKL iterative SVD solver
+@kwdef struct IterSVD
+    alg::KrylovKit.GKL = KrylovKit.GKL(; tol=1e-14, krylovdim=25)
+    lorentz_broad::Float64 = 0.0
+    alg_rrule::Union{GMRES,BiCGStab,Arnoldi} = GMRES(; tol=1e-14)
+end
+
+# Compute SVD data block-wise using KrylovKit algorithm
+function TensorKit._tsvd!(
+    t, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}, p::Real=2
+)
+    # early return
+    if isempty(blocksectors(t))
+        truncerr = zero(real(scalartype(t)))
+        return _empty_svdtensors(t)..., truncerr
+    end
+
+    Udata, Σdata, Vdata, dims = _compute_svddata!(t, alg, trunc)
+    U, S, V = _create_svdtensors(t, Udata, Σdata, Vdata, spacetype(t)(dims))
+    truncerr = trunc isa NoTruncation ? abs(zero(scalartype(t))) : norm(U * S * V - t, p)
+
+    return U, S, V, truncerr
+end
+function TensorKit._compute_svddata!(
+    t::TensorMap, alg::IterSVD, trunc::Union{NoTruncation,TruncationSpace}
+)
+    InnerProductStyle(t) === EuclideanProduct() || throw_invalid_innerproduct(:tsvd!)
+    I = sectortype(t)
+    A = storagetype(t)
+    Udata = SectorDict{I,A}()
+    Vdata = SectorDict{I,A}()
+    dims = SectorDict{I,Int}()
+    local Sdata
+    for (c, b) in blocks(t)
+        x₀ = randn(eltype(b), size(b, 1))
+        howmany = trunc isa NoTruncation ? minimum(size(b)) : blockdim(trunc.space, c)
+        S, lvecs, rvecs, info = KrylovKit.svdsolve(b, x₀, howmany, :LR, alg.alg)
+        if info.converged < howmany  # Fall back to dense SVD if not properly converged
```

Comment:
We could also consider falling back to dense SVD if `howmany / minimum(size(b)) ~ 1`, or keep this as a separate hybrid algorithm that automatically makes this choice based on some factor. I think I like having that as a separate algorithm though, because now you need to hard-code the fallback algorithm here.

Reply:
That's a great idea! I have now implemented it as an option for `IterSVD`. I think I prefer having it as an option in `IterSVD` rather than having a separate algorithm because we would have quite a bit of code duplication. Having the block-wise dense fallback is anyway a good default thing to do, I would think.